File: frontend/src/types/deployment.ts
// Mirrors backend/app/models/deployment.py and ai/schemas/deployment_specification.schema.json.
// Keep these in sync manually until an OpenAPI-generated client replaces this file.

export type RequestSource = "self_service" | "ai_assistant";
export type Environment = "dev" | "uat" | "prod";
export type ResourceType = "vpc" | "ec2" | "security-group" | "alb" | "s3" | "rds";
export type ResourceAction = "reuse" | "generate";

export interface ResourceSpec {
  type: ResourceType;
  module: string;
  version: string;
  action: ResourceAction;
  configuration: Record<string, unknown>;
}

export interface DeploymentSpecification {
  request_id: string;
  source: RequestSource;
  user_id: string;
  application: string;
  environment: Environment;
  region: string;
  resources: ResourceSpec[];
  missing_modules: string[];
  security_requirements: string[];
  approval_required: boolean;
}

export type TerraformStepStatus = "PASS" | "FAIL" | "SKIPPED";

export interface TerraformStepResult {
  step: string;
  status: TerraformStepStatus;
  detail: string;
}

export interface TerraformPlanResult {
  request_id: string;
  working_dir: string;
  steps: TerraformStepResult[];
  overall_status: TerraformStepStatus;
  generated_files: string[];
}

export interface ModuleMetadata {
  module_name: string;
  version: string;
  category: string;
  description: string;
  path: string;
  status: "approved" | "pending_review" | "deprecated" | "rejected";
  supported_environments: Environment[];
  capabilities: string[];
  security_status: "approved" | "pending_review" | "failed";
  inputs: string[];
  outputs: string[];
}

File: backend/app/models/deployment.py
"""
Pydantic models for the Deployment Specification — the canonical contract
between the UI, the AI orchestrator, and the Terraform engine.

Mirrors ai/schemas/deployment_specification.schema.json. Raw LLM output is
never used directly; it must first be parsed into (and validated as) an
AIDeploymentRecommendation, which the backend then turns into a
DeploymentSpecification.
"""
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, field_validator

from app.models.module import Environment


class RequestSource(str, Enum):
    SELF_SERVICE = "self_service"
    AI_ASSISTANT = "ai_assistant"


class ResourceType(str, Enum):
    VPC = "vpc"
    EC2 = "ec2"
    SECURITY_GROUP = "security-group"
    ALB = "alb"
    S3 = "s3"
    RDS = "rds"


class ResourceAction(str, Enum):
    REUSE = "reuse"
    GENERATE = "generate"


class ResourceSpec(BaseModel):
    type: ResourceType
    module: str
    version: str
    action: ResourceAction = ResourceAction.REUSE
    configuration: dict[str, Any] = Field(default_factory=dict)


class DeploymentSpecification(BaseModel):
    request_id: str
    source: RequestSource
    user_id: str
    application: str
    environment: Environment
    region: str
    resources: list[ResourceSpec] = Field(min_length=1)
    missing_modules: list[str] = Field(default_factory=list)
    security_requirements: list[str] = Field(default_factory=list)
    approval_required: bool = False

    @field_validator("application")
    @classmethod
    def application_must_be_slug(cls, v: str) -> str:
        if not v.replace("-", "").isalnum():
            raise ValueError("application must be an alphanumeric slug (hyphens allowed)")
        return v.lower()


# Alias for backward compatibility
DeploymentRequest = DeploymentSpecification


class DeploymentStatus(str, Enum):
    """Deployment status enumeration."""
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class DeploymentSource(str, Enum):
    """Deployment source enumeration."""
    SELF_SERVICE = "self_service"
    AI_ASSISTANT = "ai_assistant"


class AIDeploymentRecommendation(BaseModel):
    """
    The schema-validated shape an AI (Bedrock) response must conform to
    before the backend will act on it. Raw model text is never trusted —
    it is parsed into this model, and parsing failures are treated as a
    failed AI request, not silently coerced into a deployment.
    """

    request_summary: str
    environment: Environment
    region: str
    resources: list[ResourceSpec]
    missing_modules: list[str] = Field(default_factory=list)
    security_requirements: list[str] = Field(default_factory=list)
    approval_required: bool = False
